### ISSUE TITLE
Server Performance

### ISSUE THREAD

**Issue by genkimaps** [genkimaps/gis-hub#105]
@cole-fields  GIS Hub server has been unusually slow recently.  The last time this happened, in 2018, it was hackers who managed to access the server (partially) through a security hole in an old version of Solr.  

Things to check re: performance
- [x] postgres db. Check for stuck queries. Restart if needed.
- [ ] Reboot the entire server
- [x] Look for unexpected running processes (run the `htop` command)
- [x] check running services, restart as needed (postgres, redis, solr)
- [x] code in a CKAN extension that may be running slowly O(n^2)
- [x] check network connections
- [x] Solr performance?

Network connections:
`sudo netstat -tupn`

Running processes:
/usb/bin/dockerd
/usb/bin/containerd
This is odd, I don't think I have ever installed or used docker on this server. It is not part of CKAN. 

run `sudo htop -s TIME+`

redis-server is using a massive amount of processor time. Should not be happening, I think. 

Timing tests:
Using URL: https://www.gis-hub.ca/dataset
With option: `ckan.datasets_per_page = 5`
Seconds: 7, 6, 6

Disabled option: `ckan.datasets_per_page = 5`
Seconds: 47, 42, 43

Try only one dataset per page:  `ckan.datasets_per_page = 1`
Seconds: 2, 1, 1

I changed it back to `ckan.datasets_per_page = 5` after testing.

**Comment by cole-fields** [comment]
@genkimaps I suspect the slow speed is related to the ckanext-restricted extension. When i am using a session with my admin account, the site is noticeably faster than when i am using my non-admin account. I could be wrong, but it's a pattern I noticed today.

**Comment by cole-fields** [comment]
I came across a couple options for the config file we might try:

`ckan.page_cache_enabled = True`

`ckan.cache_enabled = True`

I have not added these to the .ini file.

**Comment by genkimaps** [comment]
@cole-fields  Glad you noticed this with ckanext-restricted. Have you tried disabling that in dfo.ini?

**Comment by cole-fields** [comment]
> @cole-fields Glad you noticed this with ckanext-restricted. Have you tried disabling that in dfo.ini?

I tried this today with the following results from the `/dataset` page.

| Test | Admin (seconds) | Non-Admin (seconds) |
| --- | --- | --- |
| Disable restricted from dfo.ini      | 1.6 | 5.4 |
| Enable restricted from dfo.ini      | 4.37 | 17.4 |
| Disable for loop in jinja package_item (datasets) page      | 5.6 | 17.4 |
| Enable for loop in jinja package_item (datasets) page     | 5.7 | 17.5 |

So, this suggests that the ckanext-restricted plugin is slowing things down for non-admin users. I haven't taken too close a look at the code in the extension; however, I am a bit confused because the restricted extension works to restrict access to _resources_ rather than _datasets_ and so I am curious why the slow performance when rendering the datasets list page (not the single dataset page where resources are listed and possible blocked).

**Comment by cole-fields** [comment]
Using `tail -f ckanext/restricted/restricted_action.log` from the ckanext-restricted dir, and refreshing the `/dataset' page on a non-admin account confirms that the code is checking all of the _resources_ for each dataset when rendering the `/dataset` page since it lists not authorized for each resource .

**Comment by genkimaps** [comment]
@cole-fields  It's pretty clear that the source of this problem is the restricted extension. But just to be thorough, I am looking at other things that may impact performance. 

Postgresql database:

Cleaned old DB rows with this command:
`vacuum verbose`

Look for stuck queries, there are none:

`SELECT pid, query_start, state, query FROM pg_stat_activity WHERE query != '<IDLE>' AND query NOT ILIKE '%pg_stat_activity%'  ORDER BY query_start desc;`

Restarted the following services: 
- postgresql
- redis-server
- solr (through `supervisorctl`)

Solr performance: Reloaded all Solr cores through the web interface. Not much else that can be done to improve performance on this service.

**Comment by genkimaps** [comment]
@cole-fields  Made considerable progress on this.  To make a long story short, the list view `/dataset` page uses the `package_search` API to get a list of datasets, then CKAN makes a bunch of extra API calls for *each* dataset in that list.  Inside each of those API calls, access is checked for every resource.  There was some redundant code inside the restricted extension that I cleaned up to make this faster. Was looking for a way to differentiate API calls to a specific dataset page vs. those from the dataset list view `/dataset` page. There are some things in the request context that can be used.  This improves the performance of `/dataset` from ~35 to 10 secs.

Also in the data dict, if we skip checking resource access when `include_tracking=True`, this eliminates nearly all the extra API calls, while keeping the existing behavior (web interface and API hides info from unauthorized users). This makes performance much faster, ~2 secs for `/dataset`.  

However it does open a security hole. For example this dataset has 113 resources, all restricted: https://www.gis-hub.ca/api/3/action/package_show?id=climate-projections-source-nsb

If an unauthorized user does this: https://www.gis-hub.ca/api/3/action/package_show?id=climate-projections-source-nsb&include_tracking=True

They can see everything.  There is no reason to think that a user would add this option when playing with the API, since according to the docs it has nothing to do with security: https://docs.ckan.org/en/ckan-2.7.3/api/#ckan.logic.action.get.package_show.  But still it's a bad hole and we should not leave it like this. 

This was very fiddly to debug, lot of trial and error. You can see my commits: https://github.com/genkimaps/ckanext-restricted/compare/map-preview-proxy

> We can speed things up by only checking the first resource in the list of resources to see if the user has permission to access

This sounds like a good idea, will be my next step.

**Comment by cole-fields** [comment]
@genkimaps just a note that I've merged Master into the branch the server is running (map-preview-proxy) in order to apply the nav bar dropdown functionality. 

Thanks for the work debugging this - I will take a look at the specific commits next week I think, but the performance is clearly much better. This is a considerable improvement!

**Comment by genkimaps** [comment]
@cole-fields Thanks for letting me know about the merge.  We should still talk about the implications of the include_tracking option, which as I said earlier overrides the access check: 

> If an unauthorized user does this: https://www.gis-hub.ca/api/3/action/package_show?id=climate-projections-source-nsb&include_tracking=True  They can see everything.

You merged Master into map-preview-proxy, not the other way around?

**Comment by cole-fields** [comment]
> @cole-fields Thanks for letting me know about the merge. We should still talk about the implications of the include_tracking option, which as I said earlier overrides the access check:
> 
> > If an unauthorized user does this: https://www.gis-hub.ca/api/3/action/package_show?id=climate-projections-source-nsb&include_tracking=True  They can see everything.
> 
> You merged Master into map-preview-proxy, not the other way around?

Yes, this was because I had already merged a dev branch into master that included the drop-down menu code. But the server was running on this map-preview-proxy branch which had been branched from master before the drop-down code was in it. And I figured you were still doing testing on the proxy branch so didn't want to merge it into master yet.

**Comment by cole-fields** [comment]
> More on this: the change I made works for the API. If a user makes an API request to a dataset where the first resource is private, the rest of the resources will also have sensitive info hidden. But this is not working for the web interface. When making a web request, there are a bunch of extra API calls made to other endpoints, and some of these still pick up the actual status of each resource. i.e. if the first resource is private, but other resources are open, that is how it will be displayed in the web page.

Great work on this - I'll try to make it a habit of running some reports on a regular basis and part of that will be to ensure that a dataset's resources all have the same permissions (restricted, number of allowed users, etc). 

Closing this since I believe this is complete.

**Comment by genkimaps** [comment]
@cole-fields  I closed the security hole with include_tracking: https://github.com/genkimaps/ckanext-restricted/commit/fcd25ff656f443a097af0693ce4aab1bc008049d

And tested the impact on performance. To test, I used a non-admin account that is following the MSEA org, so the dashboard page is fairly slow (all 82+ datasets have to go through the restricted check). I also tested the /dataset page.

Before  /  After
/dashboard:  12s   14s
/dataset   2.5s   2.5s
So basically no significant impact on performance. Think we can close this one finally.
